### PR TITLE
Elasticsearch: Add Top Metrics Aggregation and X-Pack support

### LINKED ISSUE #16349: Support Top Hits Aggregation for Elasticsearch
**What would you like to be added**:

Enable top_hits aggregation in the Elasticsearch queries

**Why is this needed**:

Currently, there is no way to retrieve the string fields as a metric. Most of the metrics only work for numeric fields. Top_hits will enable working with string fields.

### LINKED ISSUE #32696: Elasticsearch: Add feature toggle in datasource setting for X-Pack specific features
**What would you like to be added**:
As discussed offline with @simianhacker , we'd like to introduce a toggle in the datasource settings page to enable X-Pack specific features.

**Why is this needed**:

We still don't have any X-Pack only feature in Grafana, but we plan to add some in the near future (I.e. top metrics aggregation). 

We'd use this flag to decide whether to show or not specific features in the query editor in order to prevent users from using features in Grafana that are not supported by their ES backend.

The same setting could be used in the settings editor to show/hide other settings, for example enabling querying on frozen indices (https://github.com/grafana/grafana/issues/18277)

### PR DESCRIPTION
**Pull request by simianhacker**
**What this PR does / why we need it**:

The primary goal of this PR is to add support for `top_metrics` aggregation. There is an [open feature request](https://github.com/grafana/grafana/issues/16349) for `top_hits`, we decided to implement this instead since `top_metrics` is the recommend aggregation to replace `top_hits` for performance reasons. This PR also includes support for X-Pack only features by adding a switch to the data source (Fixes #32696). 

#### Time Series Example

image

#### X-Pack support in the datasource

image

**Which issue(s) this PR fixes**:

Fixes #32696
Fixes #16349

**Special notes for your reviewer**:

For testing this, you should use the `elastic77` source since it uses the default distribution with X-Pack enabled.

### CODE REVIEW

**Comment by Elfo404** [comment]
Would be nice to wait for https://github.com/grafana/grafana/pull/32762 and base this branch on that one if we want to move forward with it.

The reason is that https://github.com/grafana/grafana/pull/32762 will change how we handle numerical values in the query model to allow every field to be parsed as a variable, where basically internally we treat everything as a string and cast them to numbers only after variable interpolation (in the query builder).

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.test.tsx
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.test.tsx:100]: nit:

```suggestion
    it('should return top_metrics only when esVersion is equal or greater than 77', () => {
```

File: public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.test.tsx
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.test.tsx:88]: I usually try to avoid exporting things just for testing, for example in this very same file I'm testing for the field picker to be nullable in certain conditions by mounting the whole MetricEditor.

IMHO it's a more reliable way to test as it actually tests the component the same way a user would use it. what do you think?

Comment (simianhacker) [reply public/app/plugins/datasource/elasticsearch/components/QueryEditor/MetricAggregationsEditor/MetricEditor.test.tsx]: Works for me

**Comment by Elfo404** [comment]
@simianhacker  I created https://github.com/simianhacker/grafana/pull/1 to suggest some changes as the scope was too broad to use the comments here on GH.

File: public/app/plugins/datasource/elasticsearch/hooks/useFields.ts
Comment (Elfo404) [inline public/app/plugins/datasource/elasticsearch/hooks/useFields.ts:28]:
```suggestion
        // top_metrics was introduced in 7.7 where `metrics` only supported number:
        // https://www.elastic.co/guide/en/elasticsearch/reference/7.11/search-aggregations-metrics-top-metrics.html#_metrics
        // TODO: starting from 7.11 it supports ips and keywords as well:
        // https://www.elastic.co/guide/en/elasticsearch/reference/7.7/search-aggregations-metrics-top-metrics.html#_metrics
        return ['number'];
```

does it make sense?

we can add support for other fields in a follow-up PR, there's some refactoring needed here in order to select fields based on configured ES version

**Review by Elfo404 — CHANGES_REQUESTED** [review]
i'm getting a panic when testing an alert with 2 metrics:

`error="interface conversion: interface {} is nil, not float64"`
in
`/pkg/tsdb/elasticsearch/response_parser.go:642`

Query JSON Model:
```json
{
      "alias": "",
      "bucketAggs": [
        {
          "id": "2",
          "type": "date_histogram",
          "settings": {
            "interval": "auto",
            "min_doc_count": "1",
            "trimEdges": "0"
          },
          "field": "@timestamp"
        }
      ],
      "metrics": [
        {
          "id": "1",
          "settings": {
            "aggregateBy": "sum",
            "order": "desc",
            "orderBy": "@timestamp",
            "size": "1",
            "metrics": [
              "docker.cpu.user.pct",
              "docker.cpu.total.pct"
            ]
          },
          "type": "top_metrics"
        }
      ],
      "query": "",
      "refId": "A",
      "timeField": "@timestamp"
    }
```

**Comment by simianhacker** [comment]
@Elfo404  Whoops... I wasn't checking for null values. I update the test (to make it fail) then fixed the issue.

**Comment by Elfo404** [comment]
I pushed some changes to fix how we parse the response on the frontend, it seems to work as expected now both in the frontend and in the backend.

File: pkg/tsdb/elasticsearch/client/models.go
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/client/models.go:310]: How about

```suggestion
		if hasOrderBy && hasOrder {
			root["sort"] = []map[string]interface{}{
				{
					orderBy.(string): order,
				},
			}
		}
```

File: pkg/tsdb/elasticsearch/response_parser.go
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/response_parser.go:601]: `metric` is not used.

File: pkg/tsdb/elasticsearch/response_parser.go
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/response_parser.go:604]: Same comment for underscore in var names, please change :)

File: pkg/tsdb/elasticsearch/response_parser_test.go
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/response_parser_test.go:325]: Please refactor this to use `testify` instead of `convey`, as used in --> https://github.com/grafana/grafana/blob/main/pkg/api/annotations_test.go for example.

File: pkg/tsdb/elasticsearch/response_parser.go
Comment (dsotirakis) [inline pkg/tsdb/elasticsearch/response_parser.go:623]: The intermediate `metrics` array is not needed, along with this for loop. Can handle the `v.(string)` every time you get it without having to store it.

**Review by dsotirakis — CHANGES_REQUESTED** [review]

**Comment by dsotirakis** [comment]
@simianhacker 👋  and thanks for your contribution!

Please keep in mind that the `processTopMetrics(...)` func is subject to change in order to produce `Dataframes` instead of `DataTimeSeriesSlice`. We may need to revisit this if we merge the "`Series` to `Dataframes`" PR before we get to merge yours! Just FYI.

**Review by Elfo404 — APPROVED** [review]
I applied the changes suggested by @dsotirakis.

FE-wise looks and works as expected, also when used with alerting the correct data is being returned.

🚀

**Review by dsotirakis — APPROVED** [review]
Thanks for applying the changes @Elfo404!!!

Should we wait for this to be merged first? https://github.com/grafana/grafana/pull/34710

In that case we'll have to refactor the current PR a bit as well.

**Review by achatterjee-grafana — APPROVED** [review]

**Pull request merged**
